packages/ws/src/struct/WebSocketShard.ts

```diff
@@ -0,0 +1,232 @@
+import { once } from 'node:events';
+import { setTimeout } from 'node:timers';
+import { setTimeout as sleep } from 'node:timers/promises';
+import { TextDecoder } from 'node:util';
+import { AsyncQueue } from '@sapphire/async-queue';
+import { AsyncEventEmitter } from '@vladfrangu/async_event_emitter';
+import { GatewayIdentifyData, GatewayOpcodes, GatewayReceivePayload, GatewaySendPayload } from 'discord-api-types/v10';
+import { RawData, WebSocket } from 'ws';
+import type { Inflate } from 'zlib-sync';
+import type { SessionInfo } from './WebSocketManager';
+import type { IStrategy } from '../strategies/IStrategy';
+import { ImportantGatewayOpcodes } from '../utils/constants';
+import { lazy } from '../utils/utils';
+
+const getZlibSync = lazy(() => import('zlib-sync').catch(() => null));
+
+export enum WebSocketShardEvents {
+	Hello = 'hello',
+	Ready = 'ready',
+}
+
+// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
+export type WebSocketShardEventsMap = {
+	[WebSocketShardEvents.Hello]: [];
+	[WebSocketShardEvents.Ready]: [];
+};
+
+export class WebSocketShard extends AsyncEventEmitter<WebSocketShardEventsMap> {
+	private connection: WebSocket | null = null;
+
+	private readonly strategy: IStrategy;
+	private readonly id: number;
+
+	private useIdentifyCompress = false;
+
+	private inflate: Inflate | null = null;
+	private readonly textDecoder = new TextDecoder();
+
+	private isReady = false;
+
+	private sendRateLimitState = {
+		remaining: 120,
+		resetAt: Date.now(),
+	};
+
+	private readonly sendQueue = new AsyncQueue();
+
+	public constructor(strategy: IStrategy, id: number) {
+		super();
+		this.strategy = strategy;
+		this.id = id;
+	}
+
+	public async connect() {
+		// TODO(DD): Deal with already being connected
+
+		const data = await this.strategy.fetchGatewayInformation();
+
+		const { version, encoding, compression } = this.strategy.options;
+		const params = new URLSearchParams({ v: version, encoding });
+		if (compression) {
+			const zlib = await getZlibSync();
+			if (zlib) {
+				params.append('compress', compression);
+				this.inflate = new zlib.Inflate({
+					chunkSize: 65535,
+					to: 'string',
+				});
+			} else {
+				this.useIdentifyCompress = true;
+				console.warn(
+					'WebSocketShard: Compression is enabled but zlib-sync is not installed, falling back to identify compress',
+				);
+			}
+		}
+
+		this.isReady = false;
+
+		const url = `${data.url}?${params.toString()}`;
+		const connection = new WebSocket(url)
+			// eslint-disable-next-line @typescript-eslint/no-misused-promises
+			.on('message', this.onMessage.bind(this))
+			.on('error', this.onError.bind(this))
+			.on('close', this.onClose.bind(this));
```

Comment:
https://github.com/discordjs/discord.js/pull/7581#discussion_r826379489
we don't need `this` binding here.

Reply:
Thanks, but considering I haven't even got it capable of starting up a WS connection yet, this review isn't particularly helpful.

Reply:
> Thanks, but considering I haven't even got it capable of starting up a WS connection yet, this review isn't particularly helpful.

to test binding you do not need of starting up a WS connection, "this" bind is redundant here.

Reply:
This is precisely why I told you to not get hasty in the forum thread, this does NOT work without the bind, most likely because I'm using @vladfrangu's `async_event_emitter`, which does not seem to bind `this` the same as regular EE does. He should probably look into that, but for now, the binds are needed.